docs/concepts.md

```diff
@@ -1,27 +1,213 @@
+
+![ksonnet overview diagram](/docs/img/ksonnet_overview.svg)
+
+---
 
 ### Application
-TODO
+
+A ksonnet application represents a well-structured directory of Kubernetes [manifests](#manifests). This directory is autogenerated by [`ksonnet init`](docs/cli-reference/ksonnet_init.md). These manifests typically tie together in some way—collectively, they might define something like the following:
+
+<p align="center">
+<img alt="ksonnet application diagram" src="/docs/img/guestbook_app.svg" height="250px">
+</p>
+
+---
 
 ### Environment
-TODO
+
+An environment consists of four elements:
+
+| What       | Example       | Source          | How is this tracked? |
+| ---------- | ------------- | ------------- | ---------- |
+| (1) *Name* | dev | Specified at command line | A directory under `environments/` (e.g. `environments/dev/`). <br><br>**A name with slashes results in a hierarchical file structure**. For example, `us-west/staging` creates `environments/us-west/staging/`.|
+| (2) *Cluster URI* | https://cluster-name-with-hash.us-west-2.elb.amazonaws.com | Pulled from current [kubeconfig context](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#context) OR specified at command line | Inside `environments/<env-name>/spec.json` |
+| (3) *Kubernetes [namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/)* | dev | Pulled from current [kubeconfig context](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#context) OR specified at command line | Inside `environments/<env-name>/spec.json` |
+| (4) *Kubernetes API version* | version:v1.7.1 | Specified at command line, otherwise defaults to v1.7.0 | Used to generate appropriate files in `environments/<env-name>/.metadata/` based on the specified version of the Kubernetes OpenAPI spec |
+
+ksonnet allows you to deploy any particular *application* to **multiple** environments. Below is a visualization of two environments that represent different namespaces on the same cluster:
+
+![ksonnet environment diagram](/docs/img/environment.svg)
+
+If you're wondering why you might deploy a common set of manifests to different environments, here are some potential use cases:
+* Release Management (*dev* vs *test* vs *prod*)
+* Multi-AZ (*us-west-2* vs *us-east-1*)
+* Multi-cloud (*AWS* vs *GCP* vs *Azure*)
+
+---
 
 ### Component
-TODO
+
+A ksonnet *application* can be broken down into a series of discrete components:
+
+<p align="center">
+<img alt="ksonnet component diagram" src="/docs/img/component.svg" height="150px">
+</p>
+
+Components can be as simple as a single Kubernetes resource (e.g. a Deployment) or as complex as a complete logging stack (e.g. EFK). More concretely, a component corresponds to a Kubernetes manifest in `components/`. This manifest is autogenerated by the [`ks generate`](docs/cli-reference/ks_generate.md) command, and expressed in a language called [Jsonnet](#jsonnet).
```

Comment:
I don't think we should be tightly coupling components with prototypes through the generate command. I can easily see users wanting to add static manifests. Maybe `This manifest is typically autogenerated...`